### PR TITLE
Creates a Snap Checking System

### LINKED ISSUE #395: YAML Parenting only uses parented component values
PushInheritance uses a component specific code to give inheritance down to children. 

In particular this means that other prototype specific values such as placement or snapping that are not on components do not get inherited by child prototypes.

### PR DESCRIPTION
**Pull request by clusterfack**
Status: yaml works, snapgrid placement works, tryspawnentityat works, testing showed everything is fine

Creates a system that attempts to sanitize placement of objects when attempting to spawn them on snapping grids. This checks whether an object is trying to spawn specifically on a snap grid, and if it is checks against a snapflag of the type to see if other exclusive snappable objects are on that snap grid point.

Separates the spawnentityat into tryspawnentityat which checks for snappables, and forcespawnentityat which completely ignores that.
Fixes #395

### CODE REVIEW

File: SS14.Shared/Prototypes/SnapFlags.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SS14.Shared.Prototypes
+{
+    [Flags]
+    public enum SnapFlags : int
```
Comment (PJB3005) [inline SS14.Shared/Prototypes/SnapFlags.cs]: The problem is that these flags can't be extended in the slightest, so this ruins the content assembly.

I would recomment using a set of strings instead of bitflags.

File: SS14.Shared/Prototypes/IPrototype.cs [hunk 2]
```diff
@@ -16,6 +16,11 @@ public interface IPrototype
         /// Load data from the YAML mappings in the prototype files.
         /// </summary>
         void LoadFrom(YamlMappingNode node);
+
+        /// <summary>
+        /// Bitflag to hold snapping categories that this object has applied to it such as pipe/wire/wallmount
+        /// </summary>
+        SnapFlags SnapFlags { get; set; }
```
Comment (PJB3005) [inline SS14.Shared/Prototypes/IPrototype.cs]: `IPrototype` is for generic prototypes, including stuff such as recipes. This should go under `EntityPrototype`

File: SS14.Shared/Prototypes/PrototypeManager.cs [hunk 3]
```diff
@@ -60,6 +64,10 @@ public interface IPrototypeManager
         /// Syncs all inter-prototype data. Call this when operations adding new prototypes are done.
         /// </summary>
         void Resync();
+        /// <summary>
+        /// Tests whether the prototype is going to conflict with anything previously placed in this location on the snap grid
+        /// </summary>
+        bool CanSpawnAt(string EntityType, IMapGrid grid, Vector2 position);
```
Comment (PJB3005) [inline SS14.Shared/Prototypes/PrototypeManager.cs]: This functionality shouldn't be part of the prototype manager itself. It should be on entity prototypes.

File: SS14.Shared/Prototypes/EntityPrototype.cs [hunk 4]
```diff
@@ -135,6 +140,16 @@ public void LoadFrom(YamlMappingNode mapping)
             {
                 ReadPlacementProperties(placementMapping);
             }
+
+            // Reads snapping flags that this object holds that describe its properties to such as wire/pipe/wallmount, used to prevent certain stacked placement
+            if (mapping.TryGetNode<YamlSequenceNode>("snap", out var snapsequence))
```
Comment (PJB3005) [inline SS14.Shared/Prototypes/EntityPrototype.cs]: Put this under `ReadPlacementProperties` so that the syntax becomes this:
```yml
- type: entty
  ...
  placement:
    snap:
    - Pipe
    ...
```

File: SS14.Shared/GameObjects/EntityManager.cs [hunk 5]
```diff
@@ -95,6 +97,18 @@ public IEnumerable<IEntity> GetEntities(IEntityQuery query)
             return _entities.Values.Where(e => e.Match(query));
         }
 
+        public IEnumerable<IEntity> GetEntitiesAt(Vector2 position)
+        {
+            foreach (var entity in _entities.Values)
+            {
+                var transform = entity.GetComponent<ITransformComponent>();
+                if (transform.Position == position)
```
Comment (PJB3005) [inline SS14.Shared/GameObjects/EntityManager.cs]: Doing hard equality checking on floating point numbers like this is highly error prone. Allow a small range of error and use `<` and such.

File: SS14.Server/GameObjects/ServerEntityManager.cs [hunk 6]
```diff
@@ -40,12 +42,65 @@ public void SaveEntities()
...
         /// <param name="position"></param>
         /// <returns></returns>
-        public IEntity SpawnEntityAt(string EntityType, Vector2 position)
+        public bool TrySpawnEntityAt(string EntityType, IMapGrid grid, Vector2 position, out IEntity entity)
         {
-            IEntity e = SpawnEntity(EntityType);
-            e.GetComponent<TransformComponent>().Position = position;
-            e.Initialize();
-            return e;
+            var protomanager = IoCManager.Resolve<IPrototypeManager>();
```
Comment (PJB3005) [inline SS14.Server/GameObjects/ServerEntityManager.cs]: These resolves should be replaced with `[Dependency]` fields on the manager.

File: SS14.Server/GameObjects/ServerEntityManager.cs [hunk 7]
```diff
@@ -40,12 +42,65 @@ public void SaveEntities()
...
+            if(protomanager.CanSpawnAt(EntityType, grid, position))
+            {
+                entity = SpawnEntity(EntityType);
+                entity.GetComponent<TransformComponent>().Position = position;
+                entity.Initialize();
+                return true;
+            }
+            entity = new Entity(); //This is a bad
```
Comment (PJB3005) [inline SS14.Server/GameObjects/ServerEntityManager.cs]: If it's bad why are you doing it? Can't you just set the field to `null`?

File: SS14.Server/GameObjects/ServerEntityManager.cs [hunk 8]
```diff
@@ -40,12 +42,65 @@ public void SaveEntities()
...
+            return false;
+        }
+
+        /// <summary>
```
Comment (PJB3005) [inline SS14.Server/GameObjects/ServerEntityManager.cs]: These doc comments should go in the interface declaration, not the implementation.

File: SS14.Server/GameObjects/ServerEntityManager.cs [hunk 9]
```diff
@@ -40,12 +42,65 @@ public void SaveEntities()
...
+        /// </summary>
+        /// <param name="EntityType"></param>
+        /// <param name="position"></param>
+        /// <returns></returns>
+        public bool TrySpawnEntityAt(string EntityType, Vector2 position, out IEntity entity)
+        {
+            var mapmanager = IoCManager.Resolve<IMapManager>(); //TODO: wait we only have one map? this only supports one map
+            if(mapmanager.TryFindGridAt(position, out IMapGrid gridlocation))
+            {
+                bool returnvalue = TrySpawnEntityAt(EntityType, gridlocation, position, out IEntity returnoutvar);
```
Comment (PJB3005) [inline SS14.Server/GameObjects/ServerEntityManager.cs]: You can shorten this to `return TrySpawnEntityAt(EntityType, gridlocation, position, out entity);`

File: SS14.Shared/Prototypes/IPrototype.cs (hunk 2 above)
Comment (clusterfack) [reply SS14.Shared/Prototypes/IPrototype.cs]: This is true

File: SS14.Shared/Prototypes/SnapFlags.cs (hunk 1 above)
Comment (clusterfack) [reply SS14.Shared/Prototypes/SnapFlags.cs]: Ehhhh I really think strings are bad but I'm not really sure how you could run your own snapflags in content otherwise

File: SS14.Shared/GameObjects/EntityManager.cs (hunk 5 above)
Comment (clusterfack) [reply SS14.Shared/GameObjects/EntityManager.cs]: Well it works currently, if we're going to do that for checking float equality we should standardize it.

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 7 above)
Comment (clusterfack) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: No because out vars force a return value

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 6 above)
Comment (clusterfack) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: What, putting a reference to it on the entity manager with a deodency attribute or..?

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 9 above)
Comment (clusterfack) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: Cool I'll try that

File: SS14.Shared/Prototypes/EntityPrototype.cs (hunk 4 above)
Comment (clusterfack) [reply SS14.Shared/Prototypes/EntityPrototype.cs]: Sure

File: SS14.Shared/Prototypes/PrototypeManager.cs (hunk 3 above)
Comment (clusterfack) [reply SS14.Shared/Prototypes/PrototypeManager.cs]: Are we implying people are going to override entity prototypes are some point or what

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 6 above)
Comment (PJB3005) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: Yes.

https://github.com/space-wizards/space-station-14/wiki/IoC-(Inversion-of-Control)#field-injection-and-you

File: SS14.Shared/Prototypes/PrototypeManager.cs (hunk 3 above)
Comment (PJB3005) [reply SS14.Shared/Prototypes/PrototypeManager.cs]: No, but it should be part of the entity prototypes because the prototype manager is generic and designed for more than just entities.

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 7 above)
Comment (PJB3005) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: Sure about that?

https://dotnetfiddle.net/PzHqPt

Comment (RemieRichards) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: null IS a return value.
it's just a null one.

**Comment by PJB3005** [comment]
I made a PR: https://github.com/clusterfack/space-station-14/pull/1

File: SS14.Server/GameObjects/ServerEntityManager.cs (hunk 7 above)
Comment (clusterfack) [reply SS14.Server/GameObjects/ServerEntityManager.cs]: And not all types are nullable, though I guess typecasting downwards like PJB suggests shouldn't make a difference as long as people are checking the bool sanely.

**Comment by clusterfack** [comment]
Alright besides that float equals standardization function that we need to put in somewhere, this is donezo

**Comment by clusterfack** [comment]
funcode fixing #https://github.com/space-wizards/space-station-14/issues/395 incoming
